Describe the impact of ? on arrays of arrays (#1554)
dotnet/csharpstandard: standard/types.md

    ;

array_type
    : array_type nullable_type_annotation rank_specifier+
    | non_array_type rank_specifier+
    ;

> non_array_type

Reviewer: Is now the time to mention that *non_array_type* includes `System.Array` (a type required by C#) which is the supertype of all arrays?

Probably not ;-)

Author: Agreed, let's not do that here :)

    : non_array_non_nullable_type nullable_type_annotation?
    | pointer_type      // unsafe code support
    ;

non_array_non_nullable_type

>     : value_type

Reviewer: The existing grammar supported `T?[]` **iff** `T` was a value type, this PR intends to extend it to all `T`. The existing grammar supported this as *value_type* covers both nullable & non-nullable value types – the only nullable types that existed at the time. The revised grammar here provides two ways that *non_array_type* can parse `V?` as shown below – text in brackets, e.g. [`?`], following a rule indicates the input token(s) it is matching:

```
non_array_type ➜ non_array_non_nullable_type nullable_type_annotation? [`?`]
    non_array_non_nullable_type ➜ value_type
                                ➜ non_nullable_value_type  [`V`]
```

or

```
non_array_type ➜ non_array_non_nullable_type nullable_type_annotation?
    non_array_non_nullable_type ➜ value_type
                                ➜ nullable_value_type
                                ➜ non_nullable_value_type [`V`] nullable_type_annotation [`?`]
```

As there are two parses the new grammar is ambiguous, but I suspect ambiguity was not intended and this needs to be fixed in the grammar and not with a prose disambiguation rule.

Changing line 67 to reference `non_nullable_value_type` would seem to fixed the immediate issue, but I have not worked that through the PR’s total grammar changes myself.

Author: That didn't take me as long to understand as I feared - and thank you for the suggestion, which I've applied. Fortunately the name of the rule is pretty clear in terms of what it's trying to convey, too.

@BillWagner please could you check that you're happy with the change on this line as well? (`value_type` => `non_nullable_value_type`). I suspect we'll still need Nigel to run the grammar checker etc, but hopefully this will have cleared up the ambiguity.

>     : value_type

Author: ```suggestion
    : non_nullable_value_type
```

> *Note:* The types `R` and `R?` are represented by the same underlying type, `R`. A variable of that underlying type can either contain a reference to an object or be the value `null`, which indicates “no reference.” *end note*

The syntactic distinction between a *nullable reference type* and its corresponding *non-nullable reference type* enables a compiler to generate diagnostics. A compiler must allow the *nullable_type_annotation* as defined in [§8.2.1](types.md#821-general). The diagnostics must be limited to warnings. Other than in the meaning of array types, neither the presence or absence of nullable annotations, nor the state of the nullable context can change the compile time or runtime behavior of a program except for changes in any diagnostic messages generated at compile time.

> The meaning of array types is significantly impacted by the presence of *nullable_type_annotation* within an *array_type*, as described in [§17.2.1](arrays.md#1721-general).

Reviewer 2: I'd consider a different wording. The impact is that the memory layout changes. That's "significant" only if a program does anything that depends on the layout or order of multi-dimensional arrays. (Not an edge case, but not everything either).

One sugestion:

```suggestion
The memory layout of array types is impacted by the presence of *nullable_type_annotation* within an *array_type*, as described in [§17.2.1](arrays.md#1721-general).
```

Author: It's more than just the memory layout - it's what the element type of the array is. Memory layout makes it sound like too much like an implementation detail IMO.

I think the difference between declarations `string[][,] x` and `string[,][] y` is significant - which means the difference between `string[][,] x` and `string[]?[,] y` is also significant.

Reviewer: I agree with @jskeet that memory layout is not the significant impact here, I expect few users even consider it per se. The most obvious impact is that the order in which a multi-rank array[1] is indexed is changed – add a nullable annotation to a multi-rank array type and every index operation needs to be changed, I’d call that significant! And probably a sufficiently error prone undertaking to recommend avoiding it.

So I would start with Jon’s original wording. Does it need to say that this holds even if an implementation does no nullable analysis and therefore effectively ignores all nullable annotations? I think it probably needs to say that *somewhere* and here seems a reasonable place.

---
[1] Yes, under the model being described there are (I think) no multi-rank arrays and so it is technically index operations which index multiple single rank arrays, but I guess you all knew that! :-)

Author: Resolving after discussion.

> *Note:* A nullable context where both flags are disabled matches the previous standard behavior for reference types. *end note*

> The rank and element of an array type declared using *nullable_type_annotation* is not affected by the nullable context.

Reviewer: Maybe add a suitable x-ref?

> The rank and element of an array type declared using *nullable_type_annotation* is not affected by the nullable context.

Author: ```suggestion
The rank and element of an array type declared using *nullable_type_annotation* is not affected by the nullable context ([§8.9.4](types.md#894-nullable-context)).
```
